clients/lsp-yang.el

```diff
@@ -0,0 +1,126 @@
+;;; lsp-yang.el --- YANG Client settings         -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2024 Siddharth Sharma
+
+;; Author: Siddharth Sharma <siddharth.sharma@ericsson.com>
+;; Keywords: languages, yang, lsp
+
+;; This program is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; LSP support for YANG using using an external language server. Currently
+;; the supported server is:
+;;
+;; yang-lsp (yls).
+;; See https://github.com/TypeFox/yang-lsp/blob/master/docs/Settings.md
+;; for setting up the project file.
+
+;;; Code:
+
+(require 'lsp-mode)
+(require 'lsp-json)
+
+(defgroup lsp-yang nil
+  "LSP support for the YANG data modeling language using yang-lsp server."
+  :group 'lsp-yang
+  :link '(url-link "https://github.com/TypeFox/yang-lsp"))
+
+(defcustom lsp-yang-yls-version "0.7.6"
+  "yang-lsp server version to download.
+
+It has to be set before `lsp-yang.el' is loaded and it has to
+be available here: https://github.com/TypeFox/yang-lsp/releases/"
+  :type 'string
+  :group 'lsp-yang
+  :package-version '(lsp-mode . "8.0.1"))
+
+(add-to-list 'auto-mode-alist '("^yang\\.settings$" . jsonc-mode))
+
+(defcustom lsp-yang-yls-settings-schema-url
+  (format "https://raw.githubusercontent.com/TypeFox/yang-lsp/v%s/schema/yang-lsp-settings-schema.json"
+          lsp-yang-yls-version)
+  "URL for yang-lsp server settings schema"
+  :type 'string
+  :group 'lsp-yang
+  :package-version '(lsp-mode . "8.0.1"))
+
+(setq lsp-json-schemas
```

Comment:
Hmm... where does this variable come from? 🤔 This looks a bit odd to me. Usually, we don't change the variable at the root level.

Reply:
This comes from lsp-json.el in this repo. Is it better to append to the list? How does one do that for `repeat alist`?

Reply:
I guess my question is, why is the YANG lsp client required to modify JSON language server settings? 🤔 Most language server clients are isolated, so it's a bit odd to see something like this.

Reply:
The json schema association settings could be set in `init.el` or `.emacs` for user specified associations as per variable provided by `lsp-json.el`. json schema associations are not always possible to encode in the json itself due to variability in editor support and how strictly json schema is written.

The association, hence, is provided via json schema handling subsystem configuration, which is the `lsp-json.el` in lsp-mode based emacs world as I see. The validation et al under the hood happens via vscode json language server.

Now, it should be possible for "emacs extensions" which are json based config consumers, to provide schema and association (like it is possible via vscode extensions), as an addition to a list of known associations, else each user of `lsp-yang.el` for example will have to make the association themselves, or there would be a central database of association configuration, neither of which scale.

Reply:
I will remove this setting for this commit, since `setq` overrides any default settings. There should be a mechanism provided by `lsp-json.el` which allows for appending any default or further `setq`.